### ISSUE TITLE
MueLu: adding structured aggregation capabilities

### ISSUE THREAD

**Issue by lucbv** [trilinos/Trilinos#2230]
This issue tracks the progresses made toward the addition of structured aggregation capabilities in MueLu.

@rstumin @mayrmt 

@trilinos/muelu 

## Expectations
The goal is to extract,  clean up and extend the code that already exists in GeneralGeometricPFactory (GGP) to create a separate aggregation factory that can be used by TentativeP, GGP and BlackBoxP. It should allow for easier maintenance as code in each factory will be reduced.

## Current Behavior
Currently aggregation like capabilities reside in both GGP and BlackBoxP which makes it hard to upgrade the code. At the same time TentativeP is using brick aggregation which only provides limited capabilities and cannot be directly compared to GGP.

## Motivation and Context
This work is part of a bigger project to add structured grid capabilities in MueLu hopefully providing an alternative to unstructured aggregation when mesh alignment is important to the physics or numerics of a problem (line/plane smoothing).

## Definition of Done                                          
Here is a more or less chronological list of tasks that needs to happen in order to complete this work
- [x] extract code from GGP factory and create a structured aggregation factory that provides aggregates for piece-wise constant interpolation algorithm.
- [x] extend the previous work to allow uncoupled aggregation to be used with structured aggregation, currently only coupled aggregation is possible.
- [x] allow support for block structured code. This is more of a long-term goal not clearly defined yet.
- [x] add support for linear interpolation
- [ ] remove duplicated code from GGP and BlackBoxP

## Related Issues
* Is blocked by  #2577

## Additional Information
This work will require a significant amount of coding to be completed and some pieces such as block structured aggregation depend on other design decision. Therefore it is more than likely that multiple pull request will be required to complete all of the above.

**Comment by lucbv** [comment]
The structured aggregation has been merged into `develop`, this should now be accessible for the block structure work. I will start working on the decoupled version of the algorithm and create a new pull request when it's done. I might have more reviews on that one?

**Comment by lucbv** [comment]
@mayrmt started a related issue #2262 that will depends on this work.

**Comment by lucbv** [comment]
At this point we have structured aggregation for piece-wise constant and linear interpolation.
The aggregation can be performed coupled or uncoupled.
A kokkos version of the algorithm has already been put in place.
I think it is time to close this epic!
